### PR TITLE
add __state__

### LINKED ISSUE #353: Store state (built, loaded, deleted) in model
The idea here is to have a virtual field named `__state__` always generated with default value of `:built`. Loading data from the database (be it on Repo.all, Repo.insert or Repo.update) should set its status to `:loaded` while Repo.delete sets it to `:deleted`.

### PR DESCRIPTION
**Pull request by briksoftware**
Closes #353

### CODE REVIEW

File: lib/ecto/schema.ex [hunk 1]
```diff
@@ -196,6 +196,8 @@ defmodule Ecto.Schema do
       Module.register_attribute(__MODULE__, :ecto_assocs, accumulate: true)
       Module.register_attribute(__MODULE__, :ecto_raw, accumulate: true)
 
+      Module.put_attribute(__MODULE__, :struct_fields, {:__state__, :built})
```
Comment (josevalim) [inline lib/ecto/schema.ex]: Beautiful. This way it is not listed as a field anywhere, good catch!

File: test/ecto/schema_test.exs [hunk 2]
```diff
@@ -69,6 +69,10 @@ defmodule Ecto.SchemaTest do
     assert Ecto.Model.primary_key(%SchemaModel{uuid: "hello"}) == "hello"
   end
 
+  test "has __state__ attribute" do
```
Comment (josevalim) [inline test/ecto/schema_test.exs]: Can we add an assertion that checks that `SchemaModel.__schema__(:__state__) == nil`?

File: integration_test/cases/repo.exs [hunk 3]
```diff
@@ -80,16 +80,17 @@ defmodule Ecto.Integration.RepoTest do
   test "insert, update and delete" do
     post = %Post{title: "create and delete single", text: "fetch empty"}
 
-    assert %Post{} = TestRepo.insert(post)
+    assert %Post{__state__: :loaded} = TestRepo.insert(post)
```
Comment (josevalim) [inline integration_test/cases/repo.exs]: I would be more comfortable if we added those to `test/ecto/repo_test.exs` as they are not part of the integration test per se. Thoughts?

File: test/ecto/schema_test.exs (hunk 2 above)
Comment (briksoftware) [inline test/ecto/schema_test.exs]: ok, will do now

File: integration_test/cases/repo.exs (hunk 3 above)
Comment (briksoftware) [inline integration_test/cases/repo.exs]: That's where I wanted to add them, but the mock repository returns no model at all, it returns `1` apparently... but maybe I am missing something

Comment (josevalim) [inline integration_test/cases/repo.exs]: Ok, perfect, let's leave this as is!

File: test/ecto/schema_test.exs (hunk 2 above)
Comment (briksoftware) [inline test/ecto/schema_test.exs]: done

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Pull request merged**